self-core-impl/src/main/java/com/selfxdsd/core/CachingJsonResources.java

```diff
@@ -0,0 +1,241 @@
+    }
+
+    @Override
+    public Resource patch(final URI uri, final JsonValue body) {
+        return this.delegate.patch(uri, body);
+    }
+
+    @Override
+    public Resource put(final URI uri, final JsonValue body) {
+        return this.delegate.put(uri, body);
+    }
+
+    @Override
+    public Resource delete(final URI uri, final JsonValue body) {
+        return this.delegate.delete(uri, body);
+    }
+
+    /**
+     * Try to get Resource from json storage cache by URI, otherwise fetch
+     * from remote and then store the Resource.
+     * @param uri URI.
+     * @param headers Current Headers.
+     * @return Cached or remote Resource.
+     */
+    private Resource tryGetFromCache(
+        final URI uri,
+        final Supplier<Map<String, List<String>>> headers
+    ) {
+        final Resource resource;
+        String cachedEtag = this.jsonStorage.getEtag(uri);
+        if (cachedEtag != null) {
+            final Supplier<Map<String, List<String>>> updatedHeaders =
+                this.updateHeaders(headers, cachedEtag);
+            final Resource remoteResource = this.delegate
+                .get(uri, updatedHeaders);
+            final int status = remoteResource.statusCode();
+
+            if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
+                final String cachedBody = this.jsonStorage.getResourceBody(uri);
+                if (cachedBody != null) {
+                    LOG.debug(
+                        "Remote resource body for {} was not modified."
+                            + " Getting the resource body from json storage.",
+                        uri
+                    );
+                    resource = this.updateResource(remoteResource, cachedBody);
+                } else {
+                    LOG.debug(
+                        "Remote resource for {} was not modified"
+                            + " but resource body is missing from json storage."
+                            + " Getting the resource from remote.",
+                        uri
+                    );
+                    resource = this.delegate.get(uri, headers);
+                    this.tryStoreInCache(uri, resource);
+                }
+            } else {
+                LOG.debug(
+                    "Remote resource body for {} was modified or "
+                        + " has an unexpected status code.",
+                    uri
+                );
+                resource = remoteResource;
+                this.tryStoreInCache(uri, resource);
+            }
+        } else {
+            resource = this.delegate.get(uri, headers);
+            this.tryStoreInCache(uri, resource);
+        }
+        return resource;
+    }
+
+    /**
+     * Store the Resource in json storage only if Etag header is present
+     * and resource's status is HttpURLConnection.OK.
+     * @param uri URI.
+     * @param resource Resource.
+     */
+    private void tryStoreInCache(final URI uri, final Resource resource) {
```

Comment:
@criske just ``storeInCache``. Generally, try to avoid the word ``try`` in method names.